pxtlib/util.ts

```diff
@@ -1804,4 +1804,387 @@ namespace ts.pxtc.BrowserImpl {
+                            });
+                        }
+                        diff.add.push({
+                            op: 'add',
+                            path: basePath.concat(key),
+                            value: newVal
+                        });
+                    }
+                } else if (typeof oldVal !== 'object' && typeof newVal !== 'object' && oldVal !== newVal) {
+                    // Leaf nodes of same type with differing values -> replace op
+                    diff.replace.push({
+                        op: 'replace',
+                        path: basePath.concat(key),
+                        value: newVal
+                    });
+                } else if (typeof oldVal !== typeof newVal || Array.isArray(oldVal) !== Array.isArray(newVal)) {
+                    // Type changed -> replace op
+                    diff.replace.push({
+                        op: 'replace',
+                        path: basePath.concat(key),
+                        value: newVal
+                    });
+                } else {
+                    // Recurse
+                    _diff(oldVal, newVal, basePath.concat(key));
+                }
+            }
+        }
+
+        _diff(oldObj, newObj, []);
+
+        return [...diff.remove.reverse(), ...diff.replace, ...diff.add.sort((a, b) => a.path.length - b.path.length)];
+    }
+
+    /**
+     * Applies a set of JSON Patch operations to the object.
+     */
+    export function patchInPlace(obj: any, ops: PatchOperation[]): void {
+        if (!obj || typeof obj !== 'object') {
+            throw new Error("jsonPatch: Must be an object or an array.");
+        }
+        for (const op of ops) {
+            const path = op.path.slice();
+            const lastKey: (string | number) = path.pop();
+            if (lastKey == null) {
+                throw new Error("jsonPatch: missing last key");
+            }
+            let parent = obj;
+            // Find parent object of lastKey
+            let currKey = path.shift();
+            while (currKey != null) {
+                if (!(currKey in parent)) {
+                    throw new Error(`jsonPatch: missing parent element ${currKey}`);
+                }
+                parent = parent[currKey];
+                currKey = path.shift();
+            }
+            if (Array.isArray(parent) && typeof lastKey !== 'number') {
+                throw new Error(`jsonPatch: expected numeric index for array object`);
+            }
+            const ops = Array.isArray(parent) ? arrOps : objOps;
+            if (op.op === 'remove') {
+                ops.remove(parent, lastKey);
+            } else if (op.op === 'add' && !(lastKey in parent)) {
+                ops.add(parent, lastKey, op.value);
+            } else if (op.op === 'replace') {
+                ops.replace(parent, lastKey, op.value);
+            }
+        }
+    }
+}
+
+namespace ts.pxtc.jsonPatch.tests {
```

Comment:
we do have some test runners here https://github.com/microsoft/pxt/tree/master/tests (format-test or tutorial-test are pretty lightweight examples) if you eventually want to move this out of the util

Reply:
Oh nice, I'll check those out.